doc/design/attribute.md

```diff
@@ -0,0 +1,125 @@
+# Design Doc about operator attribute
+
+## background
+
+In a neural network, each operator could contain some configurable attributes. For example, a cosine similarity operator may contain an attribute named `scale`. The default cosine similarity returns a value in range [-1.0, 1.0]. But the user can set range scale manually, e.g., user set `scale=5.0`, then that cosine operator will return a value in the range [-5.0, 5.0].
+
+The configurable attributes could be various types. Some operators need `float` value to configure; some need `string` value.  We need a data structure to represent different types.
+
+Each operator contains different configurable attributes. The names of attributes are not same.  We need an associate map from attribute name to attribute value for `Operator`.
```

Comment:
In protobuf, the type of `Attributes` is a `map<string, Attribute> attrs;` But in each op class, it stores simple strong typed values as member data.

```cpp
class CosineOp {
public:
  void Init(const AttributeReader& reader) {
    scale_ = reader.Get<float>("scale")
  }

private:
  float scale_;
};
```

If we extract `scale` from protobuf map every time when we use it, it is very slow.